tests/shared/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala

```diff
@@ -1933,4 +1933,161 @@ class SyntacticSuite extends scala.meta.tests.parsers.ParseSuite {
     assertSyntax(tree, "\"\"\"\n\\\"\\\"\\\"\"\"\"")(tree)
   }
 
+  test("#2046 new line separator required before some Term.Block") {
```

Comment:
a request: could you see if it's possible to move `runTestAssert` from BaseDottySuite to its parent, ParseSuite, along with three implicit `parseXxxx` methods at the top of the class?

and then here, use it:

```
val code = "{val ..."
val syntax = s""" ... """
runTestAssert[Stat](code, Some(syntax)) {
  Term.Block(...)
}
```
that way, we not only test syntax but also that both the original code and the syntax are parsed exactly the same.

Reply:
moved it myself in #3140, perhaps you could use it.

Reply:
Ok, I'll try to use it, thanks!

Reply:
I tried to use this:
```scala
def testBlock(stat: String, expectedStr: String, needSep: Boolean): Unit = {
  def expected(sep: String): String =
    ...
  val code = s"$stat;{a}"
  val newBlock = templStat(code)
  val syntax = expected(sep = if (needSep) "\n" else "")
  runTestAssert[Stat](code, Some(syntax))(newBlock)
  ...
}
def noSep(t: String, expected: String): Unit = testBlock(t, expected, needSep = false)
noSep("type foo", "type foo")
```
But I'm having a somewhat cryptic error (at least for me). I don't see anything different in the other usages of `runTestAssert`, so I'm a little lost.
This error below has for `type foo`, but I think it's happening in all the tests:
```
Expected all trees to have non-empty `.origin`.
To fix this failure, update ScalametaParser to use `autoPos()` where the trees below got constructed.
Pro tip: you may also want to add a PositionSuite test for this tree node to verify that the position you set is correct.
=> Obtained
List(
  "Decl.Type(Nil, Type.Name(\"foo\"), Type.ParamClause(Nil), Type.Bounds(None, None))",
  "Type.Name(\"foo\")",
  "Type.ParamClause(Nil)",
  "Type.Bounds(None, None)"
)
=> Diff (- obtained, + expected)
-List(
-  "Decl.Type(Nil, Type.Name(\"foo\"), Type.ParamClause(Nil), Type.Bounds(None, None))",
-  "Type.Name(\"foo\")",
-  "Type.ParamClause(Nil)",
-  "Type.Bounds(None, None)"
-)
+Nil
    at munit.Assertions$.assertEquals(Assertions.scala:14)
    at scala.meta.tests.parsers.MoreHelpers$.requireNonEmptyOrigin(ParseSuite.scala:151)
    at scala.meta.tests.parsers.ParseSuite.runTestAssert(ParseSuite.scala:132)
    at scala.meta.tests.prettyprinters.SyntacticSuite.testBlock$1(SyntacticSuite.scala:1997)
    ...
```

Reply:
hmm, this _possibly_ means you found a bug in the parser. every tree that parser **exposes** has to have its positions within source code assigned (via an object called `origin`). it doesn't do it in all methods but in trees accessible via a public method the origin must be set.

let me check out your code and try.

Reply:
no bug :) this particular test suite intentionally resets origins, to force syntax reprinting. i moved all your tests into a new suite (and put all of them at class level, so that each case is a separate test, so that each failure is reported independently): #3143.

Reply:
> and put all of them at class level, so that each case is a separate test, so that each failure is reported independently
That is useful, the tests for this issue are quite extensive.